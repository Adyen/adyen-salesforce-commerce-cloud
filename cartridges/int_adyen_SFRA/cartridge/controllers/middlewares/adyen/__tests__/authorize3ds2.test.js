"use strict";

/* eslint-disable global-require */
var req;
var res;
var authorize3ds2;
var adyenHelper;
beforeEach(function () {
  var _require = require('../../index'),
      adyen = _require.adyen;

  adyenHelper = require('*/cartridge/scripts/util/adyenHelper');
  authorize3ds2 = adyen.authorize3ds2;
  jest.clearAllMocks();
  req = {
    form: {
      resultCode: 'IdentifyShopper',
      stateData: '{"details": {"mockeddetails":"mockedvalue"}}'
    },
    locale: {
      id: 'nl_NL'
    }
  };
  res = {
    redirect: jest.fn(),
    render: jest.fn()
  };
});
afterEach(function () {
  jest.resetModules();
});
describe('Authorize 3DS2', function () {
  it('should go to error page when authorisation fails', function () {
    var URLUtils = require('dw/web/URLUtils');

    var Logger = require('dw/system/Logger');

    authorize3ds2({}, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
    expect(Logger.error.mock.calls).toMatchSnapshot();
  });
  it('should do payments call with resultcode IdentifyShopper', function () {
    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    authorize3ds2(req, res, jest.fn());
    expect(adyenCheckout.doPaymentDetailsCall.mock.calls).toMatchSnapshot();
  });
  it('should do payments call with resultcode ChallengeShopper', function () {
    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    req.form.resultCode = 'ChallengeShopper';
    authorize3ds2(req, res, jest.fn());
    expect(adyenCheckout.doPaymentDetailsCall.mock.calls).toMatchSnapshot();
  });
  it('should redirect if form resultCode is invalid', function () {
    var URLUtils = require('dw/web/URLUtils');

    var Logger = require('dw/system/Logger');

    req.form.resultCode = 'Invalid';
    authorize3ds2(req, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
    expect(Logger.error.mock.calls).toMatchSnapshot();
  });
  it('should call fail order and redirect if result has error', function () {
    var URLUtils = require('dw/web/URLUtils');

    var OrderMgr = require('dw/order/OrderMgr');

    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    adyenCheckout.doPaymentDetailsCall.mockImplementation(function () {
      return {
        error: true
      };
    });
    authorize3ds2(req, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
    expect(OrderMgr.failOrder).toBeCalledTimes(1);
  });
  it('should redirect when result contains action', function () {
    var URLUtils = require('dw/web/URLUtils');

    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    adyenCheckout.doPaymentDetailsCall.mockImplementation(function () {
      return {
        resultCode: 'ChallengeShopper',
        action: 'mocked_action'
      };
    });
    authorize3ds2(req, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
  });
  it('should error when placing order', function () {
    var URLUtils = require('dw/web/URLUtils');

    var OrderMgr = require('dw/order/OrderMgr');

    var COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');

    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    adyenCheckout.doPaymentDetailsCall.mockImplementation(function () {
      return {
        resultCode: 'Authorised'
      };
    });
    COHelpers.placeOrder.mockImplementation(function () {
      return {
        error: true
      };
    });
    authorize3ds2(req, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
    expect(OrderMgr.failOrder).toBeCalledTimes(1);
  });
  it('should complete authorization for SFRA5', function () {
    var URLUtils = require('dw/web/URLUtils');

    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    adyenCheckout.doPaymentDetailsCall.mockImplementation(function () {
      return {
        resultCode: 'Authorised'
      };
    });
    authorize3ds2(req, res, jest.fn());
    expect(URLUtils.url.mock.calls).toMatchSnapshot();
  });
  it('should complete authorization for SFRA6', function () {
    var adyenCheckout = require('*/cartridge/scripts/adyenCheckout');

    adyenHelper.getAdyenSFRA6Compatibility.mockReturnValue(true);
    adyenCheckout.doPaymentDetailsCall.mockImplementation(function () {
      return {
        resultCode: 'Authorised'
      };
    });
    authorize3ds2(req, res, jest.fn());
    expect(res.render.mock.calls).toMatchSnapshot();
  });
});