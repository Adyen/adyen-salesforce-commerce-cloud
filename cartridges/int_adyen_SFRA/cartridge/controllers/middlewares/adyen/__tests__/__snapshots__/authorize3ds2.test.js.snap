// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorize 3DS2 should call fail order and redirect if result has error 1`] = `
Array [
  Array [
    "Checkout-Begin",
    "stage",
    "payment",
    "paymentError",
    "mocked_error.payment.not.valid",
  ],
]
`;

exports[`Authorize 3DS2 should complete authorization for SFRA6 1`] = `
Array [
  Array [
    "orderConfirmForm",
    Object {
      "orderID": "mocked_orderNo",
      "orderToken": "mocked_orderToken",
    },
  ],
]
`;

exports[`Authorize 3DS2 should complete authorization for SFRA5 1`] = `
Array [
  Array [
    "Order-Confirm",
    "ID",
    "mocked_orderNo",
    "token",
    "mocked_orderToken",
  ],
]
`;

exports[`Authorize 3DS2 should do payments call with resultcode ChallengeShopper 1`] = `
Array [
  Array [
    Object {
      "details": Object {
        "mockeddetails": "mockedvalue",
      },
      "paymentData": "mocked_adyen_payment_data",
    },
  ],
]
`;

exports[`Authorize 3DS2 should do payments call with resultcode IdentifyShopper 1`] = `
Array [
  Array [
    Object {
      "details": Object {
        "mockeddetails": "mockedvalue",
      },
      "paymentData": "mocked_adyen_payment_data",
    },
  ],
]
`;

exports[`Authorize 3DS2 should error when placing order 1`] = `
Array [
  Array [
    "Checkout-Begin",
    "stage",
    "placeOrder",
    "paymentError",
    "mocked_error.payment.not.valid",
  ],
]
`;

exports[`Authorize 3DS2 should go to error page when authorisation fails 1`] = `
Array [
  Array [
    "Error-ErrorCode",
    "err",
    "general",
  ],
]
`;

exports[`Authorize 3DS2 should go to error page when authorisation fails 2`] = `
Array [
  Array [
    "Unable to authorise 3DS2. Message: Cannot read property 'resultCode' of undefined",
  ],
]
`;

exports[`Authorize 3DS2 should redirect if form resultCode is invalid 1`] = `
Array [
  Array [
    "Checkout-Begin",
    "stage",
    "payment",
    "paymentError",
    "mocked_error.payment.not.valid",
  ],
]
`;

exports[`Authorize 3DS2 should redirect if form resultCode is invalid 2`] = `
Array [
  Array [
    "paymentDetails 3DS2 not available",
  ],
]
`;

exports[`Authorize 3DS2 should redirect when result contains action 1`] = `
Array [
  Array [
    "Adyen-Adyen3DS2",
    "orderNo",
    "mocked_orderNo",
  ],
]
`;
